### PR TITLE
Admin Crud generator

### PR DESCRIPTION
**Pull request by nsams**
depends on #1250

This is a first basic implementation of code generator for CRUD admin views that works by inspecting the gql schema.

Generated files are added to git (like we do for api, although here is not performance reason for that) - and can be reviewd in the diff.

Usage:

1. create crud-generator-config.ts:
```
import { CrudGeneratorConfig } from "@comet/cms-admin";
export default [
    {
        target: "src/products/generated",
        entityName: "Product",
    },
    {
        target: "src/news/generated",
        entityName: "News",
    },
] satisfies CrudGeneratorConfig[];
```

2. add script to package.json:
`"admin-generator": "rimraf 'src/*/generated' && comet-admin-generator generate crud-generator-config.ts",`

(and run the script)

### CODE REVIEW

File: demo/admin/src/news/generated/NewsGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            limit: dataGridProps.pageSize,
+            sort: muiGridSortToGql(dataGridProps.sortModel),
+        },
+    });
+    if (error) throw error;
+    const rows = data?.newsList.nodes ?? [];
+    const rowCount = useBufferedRowCount(data?.newsList.totalCount);
+
+    return (
+        <Box sx={{ height: `calc(100vh - var(--comet-admin-master-layout-content-top-spacing))` }}>
```
Comment (johnnyomair) [inline demo/admin/src/news/generated/NewsGrid.tsx:221]: Out of scope: We need to find a better solution for this

File: packages/admin/cms-admin/bin/comet-admin-generator.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env node
+require("ts-node").register({
+    require: ["tsconfig-paths/register"],
+});
```
Comment (johnnyomair) [inline packages/admin/cms-admin/bin/comet-admin-generator.js:4]: Why do you need this?

File: packages/admin/cms-admin/src/generator/generate-page.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+export async function writeCrudPage({ entityName, target: targetDirectory }: CrudGeneratorConfig, schema: IntrospectionQuery): Promise<void> {
+    const { classNamePlural, instanceNamePlural } = buildNameVariants(entityName);
+    const out = `
+        import { Stack, StackPage, StackSwitch } from "@comet/admin";
+        import * as React from "react";
+        import { useIntl } from "react-intl";
+        import { ${entityName}Form } from "./${entityName}Form";
+        import { ${classNamePlural}Grid } from "./${classNamePlural}Grid";
+        export function ${classNamePlural}Page(): React.ReactElement {
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-page.ts:16]:
```suggestion

        export function ${classNamePlural}Page(): React.ReactElement {
```

File: packages/admin/cms-admin/src/generator/utils/build-name-variants.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    instanceNamePlural: string;
+    fileNameSingular: string;
+    fileNamePlural: string;
+} {
+    const classNameSingular = entityName;
+    const classNamePlural = plural(entityName);
+    const instanceNameSingular = classNameSingular[0].toLocaleLowerCase() + classNameSingular.slice(1);
+    const instanceNamePlural = classNamePlural[0].toLocaleLowerCase() + classNamePlural.slice(1);
+    const fileNameSingular = instanceNameSingular.replace(/[A-Z]/g, (i) => `-${i.toLocaleLowerCase()}`);
+    const fileNamePlural = instanceNamePlural.replace(/[A-Z]/g, (i) => `-${i.toLocaleLowerCase()}`);
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/utils/build-name-variants.ts:16]: These two aren't used anywhere, please remove

File: packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import { existsSync } from "fs";
+import { IntrospectionObjectType, IntrospectionQuery } from "graphql";
+
+import { CrudGeneratorConfig } from "../types";
+
+const fallbackLibraryBlocks: { [key: string]: string } = {
+    DamImageBlock: "@comet/cms-admin",
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts:7]: What about the other blocks in the library? PixelImageBlock, InternalLinkBlock, etc.?

File: packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export function findRootBlocks({ entityName, ...generatorOptions }: CrudGeneratorConfig, schema: IntrospectionQuery) {
+    const ret = generatorOptions.rootBlocks || {};
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts:11]: Do we still need this option? You're not using it in Demo right now

File: packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { existsSync } from "fs";
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts:1]: File names in Admin code should be camelCase, you may change that for all generator files

File: packages/admin/cms-admin/src/generator/generate-form.ts [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        GQL${entityName}FormFragment,
+        GQL${entityName}FormUpdateMutation,
+        GQL${entityName}FormUpdateMutationVariables,
+        GQL${entityName}FormQuery,
+        GQL${entityName}FormQueryVariables,
+    } from "./${entityName}Form.gql.generated";
+    ${Object.entries(rootBlocks)
+        .map(([rootBlockKey, rootBlock]) => `import { ${rootBlock.name} } from "${rootBlock.import}";`)
+        .join("\n")}
+    interface FormProps {
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-form.ts:136]:
```suggestion

    interface FormProps {
```

File: packages/admin/cms-admin/src/generator/generate-form.ts [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+            return true;
+        });
+
+    const hasScope = schemaEntity.fields.some((field) => {
+        return field.name === "scope";
+    });
+
+    const outGql = `
+    import { gql } from "@apollo/client";
+    export const ${instanceEntityName}FormFragment = gql\`
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-form.ts:36]:
```suggestion

    export const ${instanceEntityName}FormFragment = gql\`
```

File: packages/admin/cms-admin/src/generator/generate-form.ts [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+    \`;
+    
+    export const ${instanceEntityName}CheckForChangesQuery = gql\`
+        query ${entityName}FormCheckForChanges($id: ID!) {
+            ${instanceEntityName}(id: $id) {
+                updatedAt
+            }
+        }
+    \`;
+    export const create${entityName}Mutation = gql\`
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-form.ts:60]:
```suggestion

    export const create${entityName}Mutation = gql\`
```

File: packages/admin/cms-admin/src/generator/generate-grid.ts [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+    ${
+        hasDeleteMutation
+            ? `const delete${entityName}Mutation = gql\`
+                mutation Delete${entityName}($id: ID!) {
+                    delete${entityName}(id: $id)
+                }
+            \`;`
+            : ""
+    }
+    ${
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-grid.ts:269]:
```suggestion

    ${
```

File: packages/admin/cms-admin/src/generator/generate-grid.ts [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+            ? `const create${entityName}Mutation = gql\`
+        mutation Create${entityName}(${hasScope ? `$scope: ${entityName}ContentScopeInput!, ` : ""}$input: ${entityName}Input!) {
+            create${entityName}(${hasScope ? `scope: $scope, ` : ""}input: $input) {
+                id
+            }
+        }
+    \`;`
+            : ""
+    }
+    function ${classNamePlural}GridToolbar() {
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-grid.ts:280]:
```suggestion

    function ${classNamePlural}GridToolbar() {
```

File: packages/admin/cms-admin/src/generator/generate-grid.ts [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+                                    await client.mutate<GQLDelete${entityName}Mutation, GQLDelete${entityName}MutationVariables>({
+                                        mutation: delete${entityName}Mutation,
+                                        variables: { id: params.row.id },
+                                    });
+                                }}
+                                `
+                                        : ""
+                                }
+                                refetchQueries={[${instanceNamePlural}Query]}
+                                // url={url}
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-grid.ts:389]:
```suggestion
```

File: demo/admin/crud-generator-config.ts [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import { CrudGeneratorConfig } from "@comet/cms-admin";
+export default [
```
Comment (johnnyomair) [inline demo/admin/crud-generator-config.ts:2]: Have you thought about instead of having a config file to extract this information from the schema as well? The `@CrudGenerator()`-Decorator could add a GraphQL comment or directive to show that the code was generated in the API and should probably be generated in the Admin as well? The `target` could be calculated using the object type name.

File: packages/admin/cms-admin/src/generator/generate-form.ts [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+        ${hasScope ? `const { scope } = useContentScope();` : ""}
+    
+        const { data, error, loading, refetch } = useQuery<GQL${entityName}FormQuery, GQL${entityName}FormQueryVariables>(${instanceEntityName}FormQuery,
+            id ? { variables: { id } } : { skip: true },
+        );
+    
+        const initialValues: Partial<FormValues> = data?.${instanceEntityName}
+            ? {
+                  ...filter<GQL${entityName}FormFragment>(${instanceEntityName}FormFragment, data.${instanceEntityName}),
+                  ${numberFields.map((field) => `${field.name}: String(data.${instanceEntityName}.${field.name}),`).join("\n")}
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/generate-form.ts:174]:
```suggestion
                  ${numberFields.length > 0 ? numberFields.map((field) => `${field.name}: String(data.${instanceEntityName}.${field.name}),`).join("\n"): ""}
```

This causes the empty line

File: demo/admin/crud-generator-config.ts (hunk 14 above)
Comment (nsams) [reply demo/admin/crud-generator-config.ts:2]: while it would be really nice to have no config file, but there is no real relation to the api generator:

- it works also for completely hand made api (at least for the parts that are compatible)
- more importantly: if you use scaffolding on api side, the `@CrudGenerator` is (currently) removed

alternative idea: *if* we would use admin generator as scaffolding only tool, we could promp for which entity to generate.

But my plan is still, that we should be able to use generation in standard cases. We will see if those cases really exist.

File: packages/admin/cms-admin/bin/comet-admin-generator.js (hunk 2 above)
Comment (nsams) [reply packages/admin/cms-admin/bin/comet-admin-generator.js:4]: needed to support config file as typescript (alternative ideas?)

File: demo/admin/src/news/generated/NewsGrid.tsx (hunk 1 above)
Comment (nsams) [reply demo/admin/src/news/generated/NewsGrid.tsx:222]: yes

File: packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts (hunk 6 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts]: The heuristics in findRootBlocks might not work correclty, that's why I left that. Do you think I should remove it?

File: packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts (hunk 5 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/utils/find-root-blocks.ts]: yes, more might be added here

- PixelImageBlock won't be used directly, always as child of DamImageBlock (correct me if I'm wrong)
- InternalLinkBlock the same (correct me if i'm wrong)
- LinkBlock is (at least in starter) a common Block

File: packages/admin/cms-admin/src/generator/generatePage.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import { IntrospectionQuery } from "graphql";
+
+import { CrudGeneratorConfig } from "./types";
+import { buildNameVariants } from "./utils/buildNameVariants";
+import { camelCaseToHumanReadable } from "./utils/camelCaseToHumanReadable";
+import { writeGenerated } from "./utils/writeGenerated";
+
+export async function writeCrudPage({ entityName, target: targetDirectory }: CrudGeneratorConfig, schema: IntrospectionQuery): Promise<void> {
```
Comment (dkarnutsch) [inline packages/admin/cms-admin/src/generator/generatePage.ts:8]: This might not be async as no Promise is returned (maybe writeGenerated needs to be awaited or returned)

**Review by dkarnutsch — APPROVED** [review]
Nice work 👏

File: packages/admin/cms-admin/src/generator/generateGrid.ts [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+    
+    const ${instanceNamePlural}Query = gql\`
+        query ${classNamePlural}Grid($offset: Int, $limit: Int${hasSort ? `, $sort: [${entityName}Sort!]` : ""}${
+        hasSearch ? `, $search: String` : ""
+    }${hasFilter ? `, $filter: ${entityName}Filter` : ""}${hasScope ? `, $scope: ${entityName}ContentScopeInput!` : ""}) {
+            ${gridQuery}(offset: $offset, limit: $limit${hasSort ? `, sort: $sort` : ""}${hasSearch ? `, search: $search` : ""}${
+        hasFilter ? `, filter: $filter` : ""
+    }${hasScope ? `, scope: $scope` : ""}) {
+                nodes {
+                    ...${entityName}sList
```
Comment (thomasdax98) [inline packages/admin/cms-admin/src/generator/generateGrid.ts:252]:
```suggestion
                    ...${entityName}List
```

File: demo/admin/src/news/generated/NewsGrid.tsx [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+        content
+        createdAt
+    }
+`;
+
+const newsQuery = gql`
+    query NewsGrid($offset: Int, $limit: Int, $sort: [NewsSort!], $search: String, $filter: NewsFilter, $scope: NewsContentScopeInput!) {
+        newsList(offset: $offset, limit: $limit, sort: $sort, search: $search, filter: $filter, scope: $scope) {
+            nodes {
+                ...NewssList
```
Comment (thomasdax98) [inline demo/admin/src/news/generated/NewsGrid.tsx:58]: Here is an "s" too much

File: demo/admin/src/news/generated/NewsForm.tsx [hunk 19]
```diff
@@ -0,0 +1,226 @@
+// This file has been generated by comet admin-generator.
+// You may choose to use this file as scaffold by moving this file out of generated folder and removing this comment.
```
Comment (thomasdax98) [inline demo/admin/src/news/generated/NewsForm.tsx:2]: How exactly is the flow if I want to use the components as a scaffold? If I remove a file from the generated folder, the CRUD generator just regenerates it the next time.

- Do I have to remove the entity from the generator config if I want to use it as a scaffold?
- Does this mean I can either auto-generate or scaffold all components of an entity (no mixture)?

- Do I have to remove the entity from the generator config?
- What would be the best way to just change one component (e.g. the form) but keep the other components generated? Or is it either all components auto-generated or all scaffolded?

Comment (nsams) [reply demo/admin/src/news/generated/NewsForm.tsx:2]: In practice I think you will remove it from the generator config and use all files as scaffolding. (no mixture)

(in theory you could have eg. grid scaffolded and grid generated)

Generating different, smaller parts that you can pick from would be a different approach I decided against, as that would be no good scaffolding (and especially in admin scaffolding is the most common use case)

File: demo/admin/src/news/generated/NewsForm.tsx [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+    const { data, error, loading, refetch } = useQuery<GQLNewsFormQuery, GQLNewsFormQueryVariables>(
+        newsFormQuery,
+        id ? { variables: { id } } : { skip: true },
+    );
+
+    const initialValues = React.useMemo<Partial<FormValues>>(
+        () =>
+            data?.news
+                ? {
+                      ...filter<GQLNewsFormFragment>(newsFormFragment, data.news),
```
[the remark this answers is not in the document]
Comment (nsams) [reply demo/admin/src/news/generated/NewsForm.tsx:78]: yes, we discussed that already somewhere else.... what was the outcome? what alternatives exist?

File: packages/admin/cms-admin/src/generator/generateGrid.ts (hunk 17 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/generateGrid.ts]: not that easy, it should be:
- ProductsList
- NewsList

File: packages/admin/cms-admin/src/generator/generatePage.ts (hunk 16 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/generatePage.ts:8]: there was an await missing....

File: demo/admin/src/news/generated/NewsGrid.tsx [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+                rowCount={rowCount}
+                columns={columns}
+                loading={loading}
+                components={{
+                    Toolbar: NewsGridToolbar,
+                }}
+            />
+        </Box>
+    );
+}
```
[the remark this answers is not in the document]
Comment (nsams) [reply demo/admin/src/news/generated/NewsGrid.tsx:237]: for my taste it's an accepable lengh, espacially as it does only a single thing.

- moving queries to own file: yes, coudl be done - you won't win too much though
- useNewsGridColumns that is getting complicated if done properly (fragment location etc)

File: demo/admin/src/news/generated/NewsForm.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (dkarnutsch) [reply demo/admin/src/news/generated/NewsForm.tsx:78]: I think this is out of scope for this MR

**Review by dkarnutsch — APPROVED** [review]

**Review by fraxachun — APPROVED** [review]

File: pnpm-lock.yaml [hunk 22]
```diff
@@ -1,9 +1,5 @@
 lockfileVersion: '6.0'
 
-settings:
-  autoInstallPeers: false
-  excludeLinksFromLockfile: false
-
```
Comment (johnnyomair) [inline pnpm-lock.yaml:6]: Why is this removed? 🤔

File: demo/admin/src/news/generated/NewsGrid.tsx [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+            </ToolbarItem>
+            <ToolbarFillSpace />
+            <ToolbarItem>
+                <GridFilterButton />
+            </ToolbarItem>
+            <ToolbarItem>
+                <Button startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add" variant="contained" color="primary">
+                    <FormattedMessage id="news.newNews" defaultMessage="New News" />
+                </Button>
+            </ToolbarItem>
```
Comment (jamesricky) [inline demo/admin/src/news/generated/NewsGrid.tsx:95]: Buttons to the right of the Toolbar should be wrapped by a single `ToolbarActions` instead of individual `ToolbarItem` components. 

 ```suggestion
            <ToolbarActions>
                <GridFilterButton />
                <Button startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add" variant="contained" color="primary">
                    <FormattedMessage id="news.newNews" defaultMessage="New News" />
                </Button>
            </ToolbarActions>
```

File: demo/admin/src/products/generated/ProductsGrid.tsx [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+            </ToolbarItem>
+            <ToolbarFillSpace />
+            <ToolbarItem>
+                <GridFilterButton />
+            </ToolbarItem>
+            <ToolbarItem>
+                <Button startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add" variant="contained" color="primary">
+                    <FormattedMessage id="product.newProduct" defaultMessage="New Product" />
+                </Button>
+            </ToolbarItem>
```
Comment (jamesricky) [inline demo/admin/src/products/generated/ProductsGrid.tsx:94]: Buttons to the right of the Toolbar should be wrapped by a single ToolbarActions instead of individual ToolbarItem components.

```suggestion
            <ToolbarActions>
                <GridFilterButton />
                <Button startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add" variant="contained" color="primary">
                    <FormattedMessage id="product.newProduct" defaultMessage="New Product" />
                </Button>
            </ToolbarActions>
```

**Review by jamesricky — CHANGES_REQUESTED** [review]

File: packages/admin/cms-admin/src/generator/generateForm.ts [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+                    <EditPageLayout>
+                        {saveConflict.dialogs}
+                        <Toolbar>
+                            <ToolbarItem>
+                                <IconButton onClick={stackApi?.goBack}>
+                                    <ArrowLeft />
+                                </IconButton>
+                            </ToolbarItem>
+                            <ToolbarTitleItem>
+                                {mode == "edit" ? values.title : <FormattedMessage id="${instanceNamePlural}.new${classNameSingular}" defaultMessage="New ${camelCaseToHumanReadable(
```
Comment (thomasdax98) [inline packages/admin/cms-admin/src/generator/generateForm.ts:265]: `values.title` shouldn't be hardcoded. It's not certain that every entity has a title field. 

This should either be "Edit ${className}" or it must be configurable via the `crud-generator-config.ts`

File: packages/admin/cms-admin/src/generator/generateForm.ts [hunk 26]
```diff
@@ -0,0 +1,358 @@
...
+    type FormValues = ${
+        numberFields.length > 0
+            ? `Omit<GQL${entityName}FormFragment, ${numberFields.map((field) => `"${field.name}"`).join(", ")}>`
+            : `GQL${entityName}FormFragment`
+    } & {
+        ${numberFields.map((field) => `${field.name}: string;`).join("\n")}
+        ${Object.keys(rootBlocks)
+            .map((rootBlockKey) => `${rootBlockKey}: BlockState<typeof rootBlocks.${rootBlockKey}>;`)
+            .join("\n")}
+    };
```
Comment (thomasdax98) [inline packages/admin/cms-admin/src/generator/generateForm.ts:158]: If there are no numberFields or rootBlocks, this results in an empty object which eslint doesn't allow. 

```ts
type FormValues = GQLCustomerFormFragment & {};
```
leads to the eslint error 
```
Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.
```

File: packages/admin/cms-admin/src/generator/generateForm.ts [hunk 27]
```diff
@@ -0,0 +1,358 @@
...
+            if (field.name === "id" || field.name === "updatedAt" || field.name === "createdAt" || field.name === "scope") return false;
+            return true;
+        })
+        .filter((field) => {
+            let type = field.type;
+            if (type.kind == "NON_NULL") type = type.ofType;
+            if (type.kind == "LIST") return false;
+            if (type.kind == "OBJECT") return false; //TODO support nested objects
+            return true;
+        });
```
Comment (thomasdax98) [inline packages/admin/cms-admin/src/generator/generateForm.ts:30]: The formFields should probably be derived from the InputType, not the ObjectType. 

For example, I explicitly told the API generator to not add `totalAmountPaid` to the input in the following entity:

```ts
@Entity()
@RootBlockEntity()
export class Order extends BaseEntity<Order, "id"> implements DocumentInterface {
    // ...

    @Property({ type: DecimalType })
    @Field(() => Float)
    @CrudField({
        input: false,
    })
    totalAmountPaid: number;

     // ...
}
```

Yet, the admin generator generates a form field for this field:

```tsx
// ...

type FormValues = Omit<GQLOrderFormFragment, "totalAmountPaid"> & {
    totalAmountPaid: string;
};

// ...

export function OrderForm({ id }: FormProps): React.ReactElement {
    
    // ...

    return (
        // ...
          <Field
              required
              fullWidth
              name="totalAmountPaid"
              component={FinalFormInput}
              type="number"
              label={<FormattedMessage id="order.totalAmountPaid" defaultMessage="Total Amount Paid" />}
          />
        // ...
    );
}
```

File: demo/admin/src/products/generated/ProductsGrid.tsx (hunk 24 above)
Comment (nsams) [reply demo/admin/src/products/generated/ProductsGrid.tsx]: changed

File: packages/admin/cms-admin/src/generator/generateForm.ts (hunk 25 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/generateForm.ts]: oops

I changed it to className for edit and add

File: packages/admin/cms-admin/src/generator/generateForm.ts (hunk 26 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/generateForm.ts]: fixed

File: packages/admin/cms-admin/src/generator/generateForm.ts (hunk 27 above)
Comment (nsams) [reply packages/admin/cms-admin/src/generator/generateForm.ts]: fixed

**Review by dkarnutsch — DISMISSED** [review]

File: demo/admin/src/news/generated/NewsGrid.tsx (hunk 1 above)
Comment (jamesricky) [reply demo/admin/src/news/generated/NewsGrid.tsx:224]: Not sure if this has been merged into `next` yet, but if it is, you can use `<MainContent />` with the `fullHeight` prop here: https://github.com/vivid-planet/comet/pull/1221

File: demo/admin/src/products/generated/ProductsGrid.tsx [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+        <Toolbar>
+            <ToolbarAutomaticTitleItem />
+            <ToolbarItem>
+                <GridToolbarQuickFilter />
+            </ToolbarItem>
+            <ToolbarFillSpace />
+            <ToolbarItem>
+                <GridFilterButton />
+            </ToolbarItem>
+            <ToolbarActions>
```
Comment (jamesricky) [inline demo/admin/src/products/generated/ProductsGrid.tsx:92]: Everything on the right side of the `Toolbar` should be wrapped with a single `ToolbarActions`, instead of an individual `ToolbarItem` like on the left side. 

```suggestion
            <ToolbarActions
                <GridFilterButton />
```

Comment (nsams) [reply demo/admin/src/products/generated/ProductsGrid.tsx]: the filter button should be left, I fixed that now.

So everything on the right (which can ATM only be the add button) is now wrapped in ToolbarActions.

**Review by jamesricky — APPROVED** [review]

**Review by johnnyomair — COMMENTED** [review]
@nsams Please resolve conflicts (probably due to new Comet version). And do you plan to resolve [this concern](https://github.com/vivid-planet/comet/pull/758#discussion_r1269429584) here or in a future PR?

**Comment by nsams** [comment]
> @nsams Please resolve conflicts (probably due to new Comet version). And do you plan to resolve [this concern](https://github.com/vivid-planet/comet/pull/758#discussion_r1269429584) here or in a future PR?

as it is merged I can add that now. Is `<MainContent fullHeight disablePadding>` correct?

**Comment by jamesricky** [comment]
> > @nsams Please resolve conflicts (probably due to new Comet version). And do you plan to resolve [this concern](https://github.com/vivid-planet/comet/pull/758#discussion_r1269429584) here or in a future PR?
> 
> as it is merged I can add that now. Is `<MainContent fullHeight disablePadding>` correct?

Yes 👍🏼

**Review by jamesricky — DISMISSED** [review]

**Review by fraxachun — DISMISSED** [review]

**Review by thomasdax98 — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]
